ipa-macros/src/parser.rs

```diff
@@ -0,0 +1,211 @@
+use crate::tree::Node;
+use quote::format_ident;
+use std::{
+    collections::{HashMap, VecDeque},
+    io::Read,
+    path::PathBuf,
+};
+
+const TARGET_CRATE: &str = "ipa";
+const STEPS_FILE_PATH: &str = "/../src/protocol/step/";
+#[cfg(not(feature = "trybuild"))]
+pub(crate) const STEPS_FILE_NAME: &str = "steps.txt";
+#[cfg(feature = "trybuild")]
+pub(crate) const STEPS_FILE_NAME: &str = "steps.test.txt";
+
+#[derive(Clone, Debug)]
+pub(crate) struct StepMetaData {
+    pub id: u16,
+    pub depth: u8,
+    pub module: String,
+    pub name: String,
+    pub path: String,
+}
+
+impl StepMetaData {
+    pub fn new(id: u16, depth: u8, module: String, name: String, path: String) -> Self {
+        Self {
+            id,
+            depth,
+            module,
+            name,
+            path,
+        }
+    }
+}
+
+/// Generate the state transition map. This is implemented as a tree where each node represents
+/// a narrowed step. The root node represents the root step, and each child node represents a
+/// narrowed step. The tree is generated by reading the steps file where each line represents a
+/// hierarchy of steps delimited by "/".
+pub(crate) fn ipa_state_transition_map() -> Node<StepMetaData> {
+    let steps = read_steps_file(STEPS_FILE_NAME)
+        .into_iter()
+        .enumerate()
+        .map(|(i, path)| {
+            let id = u16::try_from(i + 1).unwrap();
+            let path_list = path
+                .split("/")
+                .map(|s| split_step_module_and_name(s))
+                .collect::<Vec<_>>();
+            let depth = u8::try_from(path_list.len()).unwrap();
+            let (module, name) = path_list.last().unwrap();
+            // `path` is used to construct the AsRef implementation.
+            // strip the module parts from all steps to reduce the memory footprint.
+            let path = path_list
+                .iter()
+                .map(|(_, name)| name.to_owned())
+                .collect::<Vec<_>>()
+                .join("/");
+            StepMetaData::new(id, depth, module.to_owned(), name.to_owned(), path)
+        })
+        .collect::<Vec<_>>();
+
+    construct_tree(steps)
+}
+
+/// Reads the steps file and returns a vector of strings, where each string represents a line in the file.
+pub(crate) fn read_steps_file(file_path: &str) -> Vec<String> {
+    // construct the path to the steps file saved in STEPS_FILE_PATH relative to this crate's root.
+    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR").to_owned() + STEPS_FILE_PATH);
```

Comment:
This is the new code. `CARGO_MANIFEST_DIR` points to `ipa-macros`, so we go up one level, then find the file in `src/` directory ("../" part is in `STEPS_FILE_PATH`).